[inductor][triton] mark workspace args as mutated

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #134648

SplitScan makes use of a workspace arg that needs to be zeroed before it is used - then, it is used to communicate between thread blocks during the triton kernel implementation. It is mutated during during the execution of the kernel, so it should be marked as such.

Before this PR, it is not marked as mutated; AFAIK this is fine during normal execution, but during autotuning it causes problems. The workspace starts off zeroed (as expected), but during autotuning the kernel will be executed multiple times and the workspace does not get re-set between executions, resulting in incorrect data. If the data is used for indexing, then you can fail device-side asserts (and the results after the initial run (with autotuning) could be wrong). The test added in this PR repros the issue when the fix is removed.

When we mark the arg as mutated, then the arg gets cloned before autotuning, so that the arg passed to the kernel during autotuning will always be zeroed as expected.
https://github.com/pytorch/pytorch/blob/804852c1f998e6c7c58a5c7cbcb5e9063d02d54a/torch/_inductor/runtime/triton_heuristics.py#L685-L689

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

davidberard98 committed:
[inductor] pre-hooks to initialize workspace for split_scan

[ghstack-poisoned]

davidberard98 committed:
Update on "[inductor] pre-hooks to initialize workspace for split_scan"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

davidberard98 committed:
Update on "[inductor][triton] mark workspace args as mutated"


SplitScan makes use of a workspace arg that needs to be zeroed before it is used - then, it is used to communicate between thread blocks during the triton kernel implementation. It is mutated during during the execution of the kernel, so it should be marked as such.

Before this PR, it is not marked as mutated; AFAIK this is fine during normal execution, but during autotuning it causes problems. The workspace starts off zeroed (as expected), but during autotuning the kernel will be executed multiple times and the workspace does not get re-set between executions, resulting in device-side assert failures. The test added in this PR repros the issue when the fix is removed.

When we mark the arg as mutated, then the arg gets cloned before autotuning, so that the arg passed to the kernel during autotuning will always be zeroed as expected.
https://github.com/pytorch/pytorch/blob/804852c1f998e6c7c58a5c7cbcb5e9063d02d54a/torch/_inductor/runtime/triton_heuristics.py#L685-L689

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/codegen/triton.py:
```diff
@@ -2624,6 +2624,11 @@ def codegen_kernel(self, name=None):
                 mutated_args.add(self.args.output_buffers[mutation])
+
+        for argname, arg in zip(argdefs, signature):
+            if isinstance(arg, WorkspaceArg) and arg.zero_fill:
```
```suggestion
            if isinstance(arg, WorkspaceArg):
```

I don't think the zero-fill is relevant here.

davidberard98 replied:
iiuc, if zero_fill is false, then you don't actually have any requirements on the initial values in the workspace, right? in which case, do we actually need to handle it as a mutable argument (when we don't actually care about its mutations)?

peterbell10 replied:
Makes sense, worth a comment though as it's still mutating  we just don't need to restore it

davidberard98 replied:
thanks - added a comment

peterbell10 commented on test/inductor/test_torchinductor.py:
```diff
@@ -1878,6 +1878,21 @@ def fn(a, b):
+        # workspace is not zeroed.
+        def fn(lengths, data):
+            offsets = torch.cumsum(lengths, 0)
+            return data[offsets]
```
I don't think this test would ever run into issues with this.

The workspace stores partial sums that are shared between blocks. Since these don't actually change when re-running with the same arguments I don't think the result would ever be wrong. However, it would result in bad timing information as the partial results would all be available instantly and the blocks would never have to spin waiting.

tbh I'm not really sure how you could test that though. Perhaps add a verifier that runs the kernel and checks if any inputs were unexpectedly modified?

davidberard98 replied:
If you comment out the changes in codegen/triton.py and run this test, the test does fail.

During autotuning different block sizes are attempted - when you switch from testing one block size to another, you'll fail device-side assertions.

peterbell10 reviewed (commented):
Good catch, thanks!

davidberard98 committed:
add comment on "[inductor][triton] mark workspace args as mutated"


SplitScan makes use of a workspace arg that needs to be zeroed before it is used - then, it is used to communicate between thread blocks during the triton kernel implementation. It is mutated during during the execution of the kernel, so it should be marked as such.

Before this PR, it is not marked as mutated; AFAIK this is fine during normal execution, but during autotuning it causes problems. The workspace starts off zeroed (as expected), but during autotuning the kernel will be executed multiple times and the workspace does not get re-set between executions, resulting in incorrect data. If the data is used for indexing, then you can fail device-side asserts (and the results after the initial run (with autotuning) could be wrong). The test added in this PR repros the issue when the fix is removed.

When we mark the arg as mutated, then the arg gets cloned before autotuning, so that the arg passed to the kernel during autotuning will always be zeroed as expected.
https://github.com/pytorch/pytorch/blob/804852c1f998e6c7c58a5c7cbcb5e9063d02d54a/torch/_inductor/runtime/triton_heuristics.py#L685-L689

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

davidberard98 commented:
@pytorchbot merge
